### PR TITLE
Add `M-x xwwp`, update doc

### PR DESCRIPTION
**Pull request by kchanqvq**
I have this `xwwp` command in my `init.el` and find it useful so
I put it here to share.

Also updated some doc.

### CODE REVIEW

**Comment by canatella** [comment]
Nice, thank you!

File: xwwp.el [hunk 1]
```diff
@@ -161,6 +161,71 @@ Interactively, URL defaults to the string looking like a url around point."
...
+  (let ((parsed (url-generic-parse-url url)))
+    (when (url-host parsed)
+      (unless (puny-highly-restrictive-domain-p (url-host parsed))
+        (setf (url-host parsed) (puny-encode-domain (url-host parsed)))))
+    ;; When the URL is on the form "http://a/../../../g", chop off all
+    ;; the leading "/.."s.
+    (when (url-filename parsed)
+      (while (string-match "\\`/[.][.]/" (url-filename parsed))
+        (setf (url-filename parsed) (substring (url-filename parsed) 3))))
+    (setq url (url-recreate-url parsed)))
```
Comment (canatella) [inline xwwp.el:202]: Wouldn't it make sense to move this into its own function ? That way you could add a few unit test around this.

**Review by canatella — CHANGES_REQUESTED** [review]
Nice, thank you! Could you add some unit test for the xwwp--dwim-expand-url function ? Just add some tests in test/xwwp-test.el. And to run the tests: `make test`

File: xwwp.el [hunk 2]
```diff
@@ -161,6 +161,71 @@ Interactively, URL defaults to the string looking like a url around point."
       (progn (xwidget-webkit-goto-url url)
              (switch-to-buffer-other-window (xwidget-buffer (xwidget-webkit-current-session)))))))
 
+;;; Adapted from EWW code to provide a DWIM style XWWP command
+(defcustom xwwp-search-prefix "https://google.com/search?q="
+  "Prefix URL to search engine."
+  :group 'xwwp
+  :type 'string)
+
+(defun xwwp--dwim-expand-url (url)
```
Comment (canatella) [inline xwwp.el]: I don't use the `--` convention for private function anywhere. Could you use `xwwp-dwim-expand-url` here instead ?

**Comment by kchanqvq** [comment]
sure I can change naming (probably tomorrow, late now).

I don't think we need our test from those -- they are basically copied directly from eww.
Compare to eww--dwim-expand url.
This patch is kind of trivial, but I do find it convinient.

**Comment by canatella** [comment]
Is there a reason to not use `eww--dwim-expand-url` then ? Eww is distributed with emacs anyway. And agreed with the convenient, I just like to have test to avoid regression later.

**Comment by kchanqvq** [comment]
The reason I copied it and changed a bit is I find I want to have different search prefix for eww and webkit -- for eww some query url with restrictimg options need to be used.

**Comment by kchanqvq** [comment]
I think I can add some tests verifying the result of some input cases and even compare it to eww--* in case that eww ppl make their stuff better and we will know we may grab their new stuff.

It's not very likely though, those code are not updated since emacs 25 and it's using reading-string instead of better completing-read.

My next step will be to change it to completing-read so we can get a hint list like modern browsers, when using with some completion backend.

**Comment by canatella** [comment]
Lexical binding is enabled so you should be able to use:

```emacs-lisp
(let ((eww-search-prefix xwwp-search-prefix))
  (eww--dwim-expand-url "foo bar baz"))
```

**Comment by kchanqvq** [comment]
ah didn't notice this, yes.
Also formally in programming language this behavior is dynamic binding (if it's lexical binding then there should be no effect).
Will try this tomorrow.

**Comment by canatella** [comment]
Ah yes indeed, it works because `eww-search-prefix` is a `defvar` that enforce dynamic bindings, thanks for pointing that out.

**Comment by kchanqvq** [comment]
I use `eww--dwim-expand-url` now.
Any other change needed?

**Comment by kchanqvq** [comment]
Not sure if it's proper to bring it up here:

I'm now trying to implement hint based on history -- and I need to get history first.
I didn't find any WebKit JS API for global browsing history.
Seems that the only way -- which I'm currently doing -- is to modify
to `xwwp-follow-link-trigger-action` and `xwwp-follow-link-action` to get
the link and save to history, and also advice `xwidget-webkit-browse-url`.
However in this way it doesn't capture the URL if user just click on the page,
or use some other mysterious way to navigate.

Any idea?

**Review by canatella — APPROVED** [review]

**Comment by canatella** [comment]
> I use `eww--dwim-expand-url` now.
> Any other change needed?

They are two warnings that needs to be fixed for acceptance in melpa at https://github.com/canatella/xwwp/pull/8/checks?check_run_id=1004266114#step:8:12 

I guess you need to `(require 'eww)` and  `(require 'puny)`. You can run those check locally with `cask build`. Then it should be ok I guess.

**Comment by canatella** [comment]
> Not sure if it's proper to bring it up here:
> 
> I'm now trying to implement hint based on history

What do you mean by hint ? The use case is not obvious to me :)

**Comment by kchanqvq** [comment]
> > Not sure if it's proper to bring it up here:
> > I'm now trying to implement hint based on history
> 
> What do you mean by hint ? The use case is not obvious to me :)

When I type some words in the address bar of Chrome/Firefox/Safari etc they will give
completion candidates based on history URLs and titles.
It's kind of like helm in Emacs. But to do so I need to get browser history first.

Also I think browser history is useful in general, user might want to display them in a separate buffer etc.

**Comment by canatella** [comment]
Alright, I see. There is probably a C api in webkit that would give you the history but then you'd have to expose it to the lisp environment. Otherwise I guess you are stuck with this "external" history.

**Comment by canatella** [comment]
Don't forget to add the missing requires so I can merge this!

**Comment by kchanqvq** [comment]
I made the changes, also fixed the `unwind-protect` form.

**Comment by kchanqvq** [comment]
> Alright, I see. There is probably a C api in webkit that would give you the history but then you'd have to expose it to the lisp environment. Otherwise I guess you are stuck with this "external" history.

I figured out that I can advice `xwidget-webkit-callback`, then I can get full history.

Now that I'm implementing completion based on history I found that I could reuse `xwwp-follow-link-completion-***`. But currently the name suggest that they are local in package `xwwp-follow-link`.
Do you agree that I move them to `xwwp` namespace so other xwwp packages can reuse them?

Edit: actually, also less reasonable, it's easier to just `(require 'xwwp-follow-link)` in my `xwwp-history.el`.
So probably I'll just go for this.

BTW there's a commit of the ace-jump functionality in my branch. Here's a demo:
ezgif-5-48b4a484563e

**Comment by canatella** [comment]
Could you first split the ace feature in its own branch, they are unrelated and I would like to merge this PR first. I really don't understand why the unwind protect was not working before. Could you explain me how to reproduce the C-g not working for you so that I can have a better understanding on why we would need two unwind-protect ?

**Comment by kchanqvq** [comment]
Ah yes I splited the ace stuff to its own branch a while ago, currently the master branch should be clean.

There should be just one `unwind-protect`, I'll remove the duplicated one.

Originally in
```
(unwind-protect
        (condition-case nil
            (xwwp-follow-link-read xwwp-follow-link-completion-backend-instance
                                   "Link: " links
                                   (apply-partially #'xwwp-follow-link-trigger-action xwidget)
                                   (apply-partially #'xwwp-follow-link-update xwidget))
          (quit (xwwp-follow-link-cleanup xwidget)))
```
the unwind-protect form is doing nothing.
`(unwind-protect  BODYFORM UNWINDFORMS...)`, the `UNWINDFORMS` is just empty here.
So we should move the `(xwwp-follow-link-cleanup xwidget)` from the `condition-case` to the `unwind-protect`

**Comment by kchanqvq** [comment]
also I think there's a bug in xwwp-follow-link-helm and I grabbed the fix from my branch.
CI is somehow failing... look like problem from elpa?

**Comment by canatella** [comment]
Sorry to bother you, I really want to merge this but I think your `unwind-protect` changes introduces some problem in the browser it self, when I run `make test` on your branch now I get:

```
cask exec ert-runner --win
...
2020-08-24 14:28:14.313 Emacs[67744:341602] evaluateJavaScript error : A JavaScript exception occurred
2020-08-24 14:28:14.313 Emacs[67744:341602] error script=__xwidget_plus_follow_link_cleanup(null)
2020-08-24 14:28:16.311 Emacs[67744:341602] evaluateJavaScript error : A JavaScript exception occurred
2020-08-24 14:28:16.311 Emacs[67744:341602] error script=__xwidget_plus_follow_link_cleanup(null)
2020-08-24 14:28:18.814 Emacs[67744:341602] evaluateJavaScript error : A JavaScript exception occurred
2020-08-24 14:28:18.814 Emacs[67744:341602] error script=__xwidget_plus_follow_link_cleanup(null)
```
I would really like to help so if you could please explain me what is wrong with the current code in the master branch and how to reproduce the problem I could give it a shot. 

Also it would be nice to split the helm fix in its own branch/PR. I just tested and the current code is working for me and there are no issue reported by others with helm so you could explain me there how to reproduce the problem.

Thanks a lot!

**Comment by kchanqvq** [comment]
I reproduce the problem by evaluating those in `emacs -Q`
```
(require 'pacakge) (package-initialize)
(require 'helm) (helm-mode t)
(require 'xwwp) (setq xwwp-follow-link-completion-system 'helm)
```
then `M-x xwidget-webkit-browse-url https://google.com`, `M-x xwwp-follow-link`, then `C-g`.

For the helm problem, it currently focuses on the wrong link on some webpage with lots of links, e.g.
wikipedia page for emacs.

Also the `unwind-protect` fix itself should be sufficiently reasonable, and it's not related to those JS exceptions (I got those exceptions after reverting the `unwind-protect` code). Let me figure out what's wrong with those.

Edit: turns out that it is related. Cask somehow didn't rebuild the changed file during `make test`

**Comment by kchanqvq** [comment]
I just tried to use `cask emacs` instead, on your branch

There were no `C-g` cleanup problem, but the helm bug is still there.

So I think the helm fix is safe. 
`C-g` cleanup problem might depends on emacs/helm version.

**Comment by kchanqvq** [comment]
I think the reason `C-g` didn't cleanup here is that `helm` doesn't signal a `quit` error
in my environment on `C-g`, instead, it just return nil without any error. So the cleanup
in `condition-case` is not called.

Using `unwind-protect` the cleanup will always be called, but potentially extraneously.
Those calls doesn't do harm anyway. The error lines are not Emacs errors --
they are printed to stderr by WebKitGTK, and seems harmless. What do you think?

**Comment by canatella** [comment]
> I think the reason `C-g` didn't cleanup here is that `helm` doesn't signal a `quit` error
> in my environment on `C-g`, instead, it just return nil without any error. So the cleanup
> in `condition-case` is not called.

I think it might be better to make the helm backend `xwwp-follow-link-read` behave like the others by raising `quit` when it returns `nil`, does this code works with the original code:

```emacs-lisp
(cl-defmethod xwwp-follow-link-read ((backend xwwp-follow-link-completion-backend-helm) prompt collection action update-fn)
  (add-hook 'helm-after-initialize-hook (lambda ()
                                          (with-current-buffer "*helm-xwwp*"
                                            (add-hook 'helm-move-selection-after-hook update-fn nil t)))
            nil t)
  (or (helm :sources
        (helm-make-source "Xwidget Plus" 'helm-source-sync
                          :candidates collection
                          :action action
                          :filtered-candidate-transformer (lambda (candidates _)
                                                            (oset backend candidates candidates)
                                                            (funcall update-fn)
                                                            candidates))
        :prompt prompt
        :buffer "*helm-xwwp*") (signal 'quit nil)))
```
note the `or` and the `signal` call here.

> Those calls doesn't do harm anyway. The error lines are not Emacs errors --
> they are printed to stderr by WebKitGTK, and seems harmless. What do you think?

Yes they are printed by webkit but it is because we are somehow triggering an exception in the javascript code. It means we are doing something wrong. I'd like to avoid that as much as possible otherwise we might end up with lots of those errors and debugging javascript problems will become a hurdle. It is already difficult enough right now. If you can reproduce those error with the helm fix suggested above, I'll have a look at fixing them.

**Comment by kchanqvq** [comment]
Yes looks that your hack solves everything! Thanks!

**Comment by canatella** [comment]
Ok code looks nearly good, but you did not restore the `unwind-protect` that was there before, please add it back (it is there to catch possible unexpected errors). 

Then a few git history cleanup:

Could you rebase your branch to squash all the xwwp related commit in to one commit and rename that commit into something a bit more explicit like  
```
Add eww like command xwwp to search or browse

Provides a DWIM xwwp command that takes a string argument. It will display the corresponding web page if the argument is a URL. If not it will prefix the string with `xwwp-search-prefix` and visit that URL.
```

 (I try to follow https://www.git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines for writting commit messages),

Next, make sure the two helm fixes are split in their own commit, the one for the `C-g` problem with something like this for the commit message:

```
Fix xwwp-follow-link quit behaviour with Helm completion backend

When using helm and `C-g`, the Helm backend candidate function returns nil instead of propagating the `quit` signal which prevents the `xwwp-follow-link` cleanup to happen on `C-g` (the links stay highlighted) . This commit makes the Helm candidate function behave like the other ones by raising `quit` if it returns nil.
```
And another commit for the Helm link selection bug with a explicit commit message like this one.

And finally finally, could you add an `Authors` section to the readme, add myself and yourself there, something like:
```markdown
## Authors
- Damien Merenne <dam@cosinux.org>
- your name or pseudo and possibly email here.
```

**Comment by kchanqvq** [comment]
The original `(unwind-protect ...)` did nothing, as I said, since the body form were empty. Try `(unwind-protect (signal t 1))`.

If you want to "eat" any lisp error, then we should do `(condition-case nil ... (quit ...) (t nil)`, but I'm not sure if that's a good idea. I would expect a debugger and backtrace when something potentially wrong happens in an emacs package.

**Comment by canatella** [comment]
> The original `(unwind-protect ...)` did nothing, as I said, since the body form were empty. Try `(unwind-protect (signal t 1))`.
> 
> If you want to "eat" any lisp error, then we should do `(condition-case nil ... (quit ...) (t nil)`, but I'm not sure if that's a good idea. I would expect a debugger and backtrace when something potentially wrong happens in an emacs package.

Yeah, I must have added it for some reason though, but can't seem to remember

**Comment by canatella** [comment]
Ok, just tested it without, looks ok here too., lets leave it like that. Just the git history cleanup and authors and its done.

**Comment by kchanqvq** [comment]
Cleaned up git history.

CI somehow fails but `make test` passes when I run it locally.

**Comment by canatella** [comment]
Damned, it is also ok here, but the CI is running against Emacs 26, I'm installing it rn to check if I can reproduce the problem.

**Comment by canatella** [comment]
So I cannot test, I have no access to a Linux machine right now and emacs 26 on MacOS has no support for xwidget (I used the patch on 27 when developping). It seems that for some reason when running the test on CI some random page is loaded... Anyway lets merge this.

**Comment by kchanqvq** [comment]
> Yes they are printed by webkit but it is because we are somehow triggering an exception in the javascript code. It means we are doing something wrong. I'd like to avoid that as much as possible otherwise we might end up with lots of those errors and debugging javascript problems will become a hurdle. It is already difficult enough right now. If you can reproduce those error with the helm fix suggested above, I'll have a look at fixing them.

BTW, did you know that on NS emacs you can right click > inspect element to get a full debug console? I didn't know this until I just read `nsxwidget.m` source.

**Pull request merged**